oprf/oprf.go

```diff
@@ -0,0 +1,277 @@
+}
+
+//GenerateKeyPair generates a KeyPair in accordance with the group.
+func GenerateKeyPair(suite *group.Ciphersuite) *KeyPair {
+	privK := suite.RandomScalar()
+	pubK := suite.ScalarMultBase(privK)
+
+	return &KeyPair{pubK, privK}
+}
+
+func assignKeyPair(suite *group.Ciphersuite, privK, pubK []byte) (*KeyPair, error) {
+	kp := &KeyPair{}
+	err := kp.Deserialize(suite, privK, pubK)
+	if err != nil {
+		return nil, err
+	}
+
+	return kp, nil
+}
+
+func suiteFromID(id SuiteID, ctx []byte) (*group.Ciphersuite, error) {
+	var err error
+	var suite *group.Ciphersuite
+
+	switch id {
+	case OPRFP256:
+		suite, err = group.NewSuite("P-256", ctx)
+	case OPRFP384:
+		suite, err = group.NewSuite("P-384", ctx)
+	case OPRFP521:
+		suite, err = group.NewSuite("P-521", ctx)
+	default:
+		return suite, ErrUnsupportedGroup
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return suite, err
+}
+
+// NewServer creates a new instantiation of a Server. It can create
+// a server with existing keys or use pre-generated keys.
+func NewServer(id SuiteID, privK, pubK []byte) (*Server, error) {
+	ctx := generateCtx(id)
+
+	suite, err := suiteFromID(id, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var keyPair *KeyPair
+	if privK == nil || pubK == nil {
+		keyPair = GenerateKeyPair(suite)
+	} else {
+		keyPair, err = assignKeyPair(suite, privK, pubK)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &Server{
+			suite: suite,
+			ctx:   ctx,
+			Kp:    keyPair},
+		nil
+}
+
+// Evaluate blindly signs a client token.
+func (s *Server) Evaluate(b BlindToken) (*Evaluation, error) {
+	p := group.NewElement(s.suite)
+	err := p.Deserialize(b)
+	if err != nil {
+		return nil, err
+	}
+
+	z := p.ScalarMult(s.Kp.PrivK)
+	ser := z.Serialize()
+
+	return &Evaluation{ser}, nil
+}
+
+// FullEvaluate performs a full evaluation at the server side.
+func (s *Server) FullEvaluate(in, info []byte) ([]byte, error) {
+	p, err := s.suite.HashToGroup(in)
+	if err != nil {
+		return nil, err
+	}
+
+	t := p.ScalarMult(s.Kp.PrivK)
+	iToken := t.Serialize()
+
+	h := group.FinalizeHash(s.suite, in, iToken, info, s.ctx)
+
+	return h, nil
+}
+
+// VerifyFinalize verifies the evaluation.
+func (s *Server) VerifyFinalize(in, info, out []byte) bool {
+	p, err := s.suite.HashToGroup(in)
+	if err != nil {
+		return false
+	}
+
+	el := p.Serialize()
+
+	e, err := s.Evaluate(el)
+	if err != nil {
+		return false
+	}
+
+	h := group.FinalizeHash(s.suite, in, e.element, info, s.ctx)
+	return subtle.ConstantTimeCompare(h, out) == 1
+}
+
+// NewClient creates a new instantiation of a Client.
+func NewClient(id SuiteID) (*Client, error) {
+	ctx := generateCtx(id)
+
+	suite, err := suiteFromID(id, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &Client{}
+	client.suite = suite
+	client.ctx = ctx
+
+	return &Client{
+		suite: suite,
+		ctx:   ctx,
+	}, nil
+}
+
+// Request generates a token and its blinded version.
+func (c *Client) Request(in []byte) (*Token, BlindToken, error) {
```

Comment:
Hmm, is it best that the `Request` function output the `Token` and `BlindToken` separately? This seems like an easy way for applications to misuse the API, e.g., by using the blindToken from one Request() call in a different Finalize() call. I think we should encapsulate these in a new type, and then make `Finalize` a function on that type. For example:

```go
type ClientRequest struct {
   input []byte
   token *Token
   blindToken BlindToken
}

func (c *Client) Request(in []byte) (ClientRequest, error) {
   // return a ClientRequest
}

func (r ClientRequest) Finalize(e * Evaluation) ([]byte, error) {
  // compute and return OPRF output
}
```

Also, why does `Finalize` need to return the `IssuedToken`? That doesn't seem like useful output for the protocol.

Reply:
I like this idea, also we can extend the `ClientRequest` structure to support a list of tokens per request.
This is more useful when implementing the verifiable mode.

> Also, why does Finalize need to return the IssuedToken? 

I don't think so, unless for some reason the user wants to store them.

Reply:
Yeah, I think the Finalize output should only be the OPRF output, and nothing more.

Reply:
Mmm.. Finalize does not ask as a parameter for a the BlindToken, and the BlindToken is of a different type than the Token (one is an struct and the other is a byte array), so there is no way that the BlindToken can be used as a parameter for `Finalize`. 

```
type ClientRequest struct {
   input []byte
   token *Token
   blindToken BlindToken
}
```

Why you need the input if that is already part of the Token struct?

The struct defined here (`ClientRequest`) does not have access to the Ciphersuite (needed to access the type of hash defined by the ciphersuite, needed to do the finalizeHash call), which is accessible in the Client struct. Those parameters will need to be copied form the `Client` struct toe the `ClientRequest` struct. 

I don't really follow the misuse case.

Reply:
Should be added now, though.

Reply:
We can always change what goes in `ClientRequest` :-) To elaborate on the misuse, consider the following: 

```go
inputA := []byte("inputA")
inputB := []byte("inputB")

tokenA, blindTokenA, errA := c.Request(inputA)
tokenB, blindTokenB, errB := c.Request(inputB)

// ... send off blindTokenA and blindTokenB to get evaluationA and evaluationB

outputA, errA := c.Finalize(tokenA, evaluationB, "info")
outputB, errB := c.Finalize(tokenB, evaluationA, "info") // oops -- I swapped these things around!
```

```go
inputA := []byte("inputA")
inputB := []byte("inputB")

requestA, errA := c.Request(inputA)
requestB, errB := c.Request(inputB)

// ... send off blindTokenA (from requestA) and blindTokenB (from requestB) to get evaluationA and evaluationB

outputA, errA := requestA.Finalize(evaluationA, "info")
outputB, errB := requestB.Finalize(evaluationB, "info")
```

The potential misuse doesn't go away (obviously), but now the caller only needs to match one request to one evaluation, which seems simpler to me. It also lets us batch up multiple requests in a single object, which is nice.

Reply:
Sure. Added that way.